### PR TITLE
Add configuration option to disable Express middleware span instrumentation

### PR DESCRIPTION
**Pull request by ericmustin**
### What does this PR do?

This PR addresses the use case introduced in [Issue-765](https://github.com/DataDog/dd-trace-js/issues/764) for when a user wants to automatically disable the express plugins instrumentation of middleware via configuration option.

The current behavior is that all express middleware is automatically instrumented when plugins are automatically enabled at tracer `.init` call, or when a user invokes `.use('express')`. In some cases this can add unwanted `express.middleware` spans that the user doesn't wish to instrument as it causes an excessively high number of spans with incrementally little value to the particular user, or performance overhead for instrumenting these additional spans. While this can be filtered out of Datadog's UI, users have also requested to disable middleware instrumentation for express.

This PR introduces a `disableMiddleware` boolean configuration option for the express integration. It's written so that it could be eventually extended to other web frameworks with a high number of low-information middleware spans, such as koa or hapi, but at this time there is only support and testing for express.

usage would be:

- configuration option: 

```
	tracer = require('dd-trace')
	tracer.use('express', { disableMiddleware: true})
	tracer.init()
```

And you can see the difference in output in the UI here, comparing how the traces look with the middleware disabled vs default (enabled)

default (enabled):

default framegraph
default span list

disabled ({disableMiddleware: true}): 

middleware disabled flamegraph
middleware disabled span list

### Motivation

user request

### Plugin Checklist

- [x] Unit tests.
- [x] TypeScript [definitions][1].
- [ ] TypeScript [tests][2].
- [x] API [documentation][3].
- [ ] CircleCI [jobs/workflows][4].
- [ ] Plugin is [exported][5].

[1]: https://github.com/DataDog/dd-trace-js/blob/master/index.d.ts
[2]: https://github.com/DataDog/dd-trace-js/blob/master/docs/test.ts
[3]: https://github.com/DataDog/dd-trace-js/blob/master/docs/API.md
[4]: https://github.com/DataDog/dd-trace-js/blob/master/.circleci/config.yml
[5]: https://github.com/DataDog/dd-trace-js/blob/master/packages/dd-trace/src/plugins/index.js

### Additional Notes

This is just a first pass, let me know if my approach is appropriate here, as i'm doing a bit of a lower level change (to the general `router` patching, and `util/web.js` helper functions) for something that's currently only available to express, instead of all plugins that extend httpserver. It should be relatively straightforward to add support for this to other extensions of httpserver such as koa, connect, or restify, I just don't have much experience with those libraries and wanted to knockout the highest impact plugin (express) that this request seems to come up for most frequently. I am a little concerned there's some edge cases introduced by this option where an error would not get captured (perhaps it should sit in `experimental` for a few versions until there's more confidence around it) but the tests seem to pass fine for this scenario.

### CODE REVIEW

**Comment by rochdev** [comment]
> I am a little concerned there's some edge cases introduced by this option where an error would not get captured

It should be pretty straightforward to find any error that are manually added to the span in the plugins and make sure the active span is used instead of a specific span.

**Comment by ericmustin** [comment]
@rochdev re: 

>

>  I would prefer an approach that could apply to every plugin with the same syntax. For example:
> 
> tracer.use('express', {
>   enabled: {
>     request: true,
>     middleware: false
>   }
> })
> What do you think?

So i think we're just discussing web server plugins here, but definitely having `middleware` as a general option is feasible, though i have to get a little more familiar with how/if we are instrumenting middlewares beside the ones that use the very straightforward `web.wrapMiddleware` utility function. Looks like just koa, connect, and express at the moment use `web.wrapMiddleware`.

for those three I think it would be pretty easy to extend/generalize this PR to offer that syntax across those three, but I'm not sure how straightforward it would be for the other web servers. 

As far as offering the same sort've general config syntax for the `request` spans, i think the scope of that work would be much bigger as it seems like there's a lot more variability around how we instrument the `<integration_name>.request` span and more broadly I'm not sure if that's an option that would be particularly useful for users since that's where a majority of these plugins value lives?

**Comment by rochdev** [comment]
The request option was just an example to demonstrate that eventually I expect to have support for choosing only the spans you want in every integration. For example, maybe some users don't need `parse` and `validate` spans for GraphQL and would like to disable that. For now we can implement only `middleware`.

**Comment by ericmustin** [comment]
Ok makes sense, I can work on a general `middleware` approach, should be straightforward

**Comment by ericmustin** [comment]
@rochdev I have bungled the tests here and am going to try to clean this up this week, however I abstracted this out to the other http we server integrations that instrument middlewares (koa, connect would work currently)

**Comment by ericmustin** [comment]
hey there, @rochdev , hope all is well. So I've updated this PR so that a `middleware` option has been abstracted a bit lower down to the `httpServer` level, making it a general configuration option for any web server frameworks instrumentation that extend `httpServer`. Implicitly there's only three of these at the moment which actually instrument their middleware (`express,koa,connect`) as far as I can tell, so this is functionally just a no-op config option for any of the other web server framework instrumentation at the moment. Not sure if that ought to be documented anywhere.

I think this works better and actually this approach opens up a lot of configurability down the road, for example, users who want to disable instrumentation of _certain_ middleware there could be some sort of regex or whitelist/blacklist type functionality, etc.

Let me know what you think of the implementation, and specifically I'm still a little unsure on one bit here which is the helper function `wrapMiddlewareErrors` to ensure that if there's an unhandled error in a middleware, our instrumention appends the error msg, stack trace, etc to the active span (which would be `express.request` or `koa.request` or what have you), as there's no middleware span to append it to. I've added some tests here for this, but A. wasn't totally sure how to handle attaching the relevant metadata from the middleware error (I basically just wrap the original middleware function in a try/catch block and then grab the relevant info and  then rethrow the error), and also B. this doesn't appear to play nicely with the status code validator option (which is causing one expressJS test to fail), so I have to figure that bit out still, but wanted to know your thoughts here.

**Comment by manelpb** [comment]
any change we could get this merged soon?

**Comment by ericmustin** [comment]
just to update, i've finally had some free cycles during the holidays to fix that last breaking spec....still needs a thorough review (after 🎅 🎄 of course 😉 ) per the details/questions i noted above in last comment

File: packages/datadog-plugin-connect/test/index.spec.js [hunk 1]
```diff
@@ -504,82 +505,234 @@ describe('Plugin', () => {
...
+                  .get(`http://localhost:${port}/app/user/1`)
+                  .catch(done)
+              })
             })
           })
-        })
 
-        it('should include specified headers in metadata', done => {
-          const app = connect()
+          it('should not activate a scope per middleware', done => {
```
Comment (rochdev) [inline packages/datadog-plugin-connect/test/index.spec.js]: A scope should still be activated, but it should be the one of the request instead.

File: packages/datadog-plugin-connect/test/index.spec.js [hunk 2]
```diff
@@ -504,82 +505,234 @@ describe('Plugin', () => {
           connect = require(`../../../versions/connect@${version}`).get()
         })
 
-        it('should be configured with the correct service name', done => {
-          const app = connect()
+        describe('with middleware disabled', () => {
```
Comment (rochdev) [inline packages/datadog-plugin-connect/test/index.spec.js]: This tests more than that.

File: packages/datadog-plugin-express/test/index.spec.js [hunk 3]
```diff
@@ -714,7 +715,7 @@ describe('Plugin', () => {
 
           app.get('/user', (req, res) => {
             res.statusCode = 400
-            throw new Error('boom')
+            res.status(400).send()
```
Comment (rochdev) [inline packages/datadog-plugin-express/test/index.spec.js]: Why this change?

File: packages/datadog-plugin-connect/src/index.js [hunk 4]
```diff
@@ -38,7 +40,7 @@ function createWrapUse () {
       const layer = this.stack[index]
 
       if (layer && layer.handle) {
-        this.stack[index].handle = wrapLayerHandle(layer)
+        this.stack[index].handle = wrapLayerHandle(layer, config)
```
Comment (rochdev) [inline packages/datadog-plugin-connect/src/index.js]: Would it require less passing around to provide it to `web.patch` instead and store it on the request? It would also make the config available to other web helpers if needed.

Comment (ericmustin) [reply packages/datadog-plugin-connect/src/index.js]: I tend to like the idea of explicitly passing the config down to where it's required rather than relying on some other bit of instrumentation to make it available? It also make it a bit easier for me to follow logically. If you feel like this is a blocker I'm open to revamping it, it just may take me a bit of time to refactor everything.

File: packages/datadog-plugin-connect/test/index.spec.js (hunk 2 above)
Comment (ericmustin) [reply packages/datadog-plugin-connect/test/index.spec.js]: good point. I've updated these tests so that the testing of the other parts of the config is separate from testing the middleware functionality

File: packages/datadog-plugin-connect/test/index.spec.js (hunk 1 above)
Comment (ericmustin) [reply packages/datadog-plugin-connect/test/index.spec.js]: Thank you for highlighting this. I've updated the underlying middleware instrumentation so that it now both leverages `web.reactivate` for the request span, and also captures any errors thrown by the middleware, and updated the tests accordingly to match this behavior (that the middleware is exposed to the request scope, not their own scope and not a null scope) . lmk what you think

File: packages/datadog-plugin-express/test/index.spec.js (hunk 3 above)
Comment (ericmustin) [reply packages/datadog-plugin-express/test/index.spec.js]: Explicitly throwing an error in the middleware, when middleware instrumentation is disabled, gets picked up regardless of the statusCode due to the additional logic introduced by this PR.

I had considered also doing some work within the `reactivateAndWrapMiddlewareErrors` (yea, the name is gross, sorry, open to suggestions) method so that it only _conditionally_ marks a span as an error based on `validateStatus` but it's not clear to me that that the response's statusCode will always be set intentionally within the middleware. 

As this test is meant to test status Code handling it made sense to me to just simplify this test. This seems to be the only test this sortve edge case really touches but I suppose I am introducing a slightly new behavior here in the situation where a user has both disabled middleware instrumentation and wants to return a non-error status code while throwing an error in the middleware? 

Lmk what you think, I also may have explained that poorly so please lmk if i can clarify anything.

Comment (rochdev) [reply packages/datadog-plugin-express/test/index.spec.js]: Still not sure I understand. The reason for throwing in this test was that throwing is a different code path than responding. In any case, the status code should be handled properly at the HTTP level since if the error is unhandled it will reach the request, and if it's handled then it will be the user decision.

File: packages/datadog-plugin-connect/src/index.js (hunk 4 above)
Comment (rochdev) [reply packages/datadog-plugin-connect/src/index.js]: One of the main goals of the web util is to remove as much tracing code as possible from the integrations and instead attach everything to the request. Considering that, I think it would be better to set the config as soon as possible so that it doesn't need to be passed again everywhere that it's needed.

File: packages/datadog-plugin-express/test/index.spec.js (hunk 3 above)
Comment (ericmustin) [reply packages/datadog-plugin-express/test/index.spec.js]: Ok, I think I misunderstood something here. In the other test case for `should only handle errors for configured status codes` There is still an error being marked, however it's on the middleware span. In this case we will still need to mark an error, but the request span is the active scope so it needs to be marked there vs the middleware (which we are not tracking separately with middleware tracing disabled). In this case I've also updated the spec test to be explicit about the behavior `should mark middleware errors regardless of status codes configuration`

File: packages/datadog-plugin-connect/src/index.js (hunk 4 above)
Comment (ericmustin) [reply packages/datadog-plugin-connect/src/index.js]: alright, fair enough, will try to knock out this change

Comment (ericmustin) [reply packages/datadog-plugin-connect/src/index.js]: well, it would appear i am obtuse, as the config already exists on `req._datadog` by the time the request is passed into the wrapMiddleware function.

with `web.instrument`, immediately after .patch is called [here](https://github.com/DataDog/dd-trace-js/blob/564f0782327620243a78ee5a544248a37458acd9/packages/dd-trace/src/plugins/util/web.js#L52-L54) we invoke `startSpan` which sets `_datadog.config` [here](https://github.com/DataDog/dd-trace-js/blob/564f0782327620243a78ee5a544248a37458acd9/packages/dd-trace/src/plugins/util/web.js#L211)

So all this passing down of `config` appears to have been totally unnecessary on my part. That being said, I figure it can't hurt to at least explicitly define an empty config object in the  `patch` call on the request's `_datadog` property, which would make any later calls to check `req._datadog.config` a bit more defensive since worst case the config value would be undefined. What do you think? I think this is much cleaner now for sure

Comment (ericmustin) [reply packages/datadog-plugin-connect/src/index.js]: i suppose i could also just explicitly pass in config to `.patch` and set it within `patch`, if that just simplifies things even more

File: packages/datadog-plugin-express/test/index.spec.js [hunk 5]
```diff
@@ -986,6 +985,207 @@ describe('Plugin', () => {
...
+
+        after(() => {
+          return agent.close()
+        })
+
+        beforeEach(() => {
+          express = require(`../../../versions/express@${version}`).get()
+        })
+
+        it('should not activate a scope per middleware', done => {
```
Comment (rochdev) [inline packages/datadog-plugin-express/test/index.spec.js]: This test should explicitly try to change the scope in the previous middleware to ensure the next one properly reactivates the scope of the request and that it's not just propagated because of the execution context. This is to make sure you get the right scope when the previous middleware lost it somehow.

I assume this also applies to other plugin tests.

File: packages/datadog-plugin-express/test/index.spec.js [hunk 6]
```diff
@@ -986,6 +985,207 @@ describe('Plugin', () => {
...
+
+          getPort().then(port => {
+            appListener = app.listen(port, 'localhost', () => {
+              axios.get(`http://localhost:${port}/user`)
+                .catch(done)
+            })
+          })
+        })
+
+        it('should not activate a span for every middleware on a route with middleware disabled ', done => {
```
Comment (rochdev) [inline packages/datadog-plugin-express/test/index.spec.js]: If I understand the test properly, the title is misleading. Basically this is checking that middleware are not instrumented?

File: packages/dd-trace/src/plugins/util/web.js [hunk 7]
```diff
@@ -102,6 +106,24 @@ const web = {
...
   },
 
+  // catch errors and apply to active span
+  reactivateAndWrapMiddlewareErrors (fn, req) {
+    try {
+      return this.reactivate(req, fn)
+    } catch (e) {
+      const activeSpan = this.active(req)
+      if (activeSpan) {
+        activeSpan.addTags({
```
Comment (rochdev) [inline packages/dd-trace/src/plugins/util/web.js]: This could be `setTag('error', e)`.

File: packages/dd-trace/src/plugins/util/web.js [hunk 8]
```diff
@@ -102,6 +106,24 @@ const web = {
     return tracer.scope().activate(span, fn)
   },
 
+  // catch errors and apply to active span
+  reactivateAndWrapMiddlewareErrors (fn, req) {
```
Comment (rochdev) [inline packages/dd-trace/src/plugins/util/web.js]: Why is this needed? The scope manager should already be doing that.

File: packages/datadog-plugin-connect/src/index.js (hunk 4 above)
Comment (rochdev) [reply packages/datadog-plugin-connect/src/index.js]: Do you mean for `getMiddlewareSetting`? If that's the case, I think for now whatever is simplest. Eventually we should probably do this in a safe and consistent way.

**Comment by rochdev** [comment]
A few questions as I'm still trying to understand every part of the change, but otherwise I think this is very close to being done 👍

File: packages/datadog-plugin-express/test/index.spec.js (hunk 5 above)
Comment (ericmustin) [reply packages/datadog-plugin-express/test/index.spec.js]: makes sense, updated these tests(along with those in connect and koa) to more explicitly test that.

File: packages/dd-trace/src/plugins/util/web.js (hunk 7 above)
Comment (ericmustin) [reply packages/dd-trace/src/plugins/util/web.js]: passing in the whole object was causing error.type to not get set (since it's set to `e.name`). Are `type` and `name` interchangeable values for errors? happy to update to the suggested change and update related test as well if so, or happy to leave as is if not.

File: packages/datadog-plugin-connect/src/index.js (hunk 4 above)
Comment (ericmustin) [reply packages/datadog-plugin-connect/src/index.js]: no not particularly, `getMiddlewareSetting` is for ensuring the config.middleware is normalized and if not set, defaults to `true`. Similar to `getStatusValidator`

What i was referring to is that I was passing config down all the way into every invocation of `web.wrapMiddleware` like [here](https://github.com/DataDog/dd-trace-js/blob/5348eaad6f1a86d57b87ffd3de6397feb1dcb417/packages/datadog-plugin-router/src/index.js#L116) , but since `web.instrument` already handles setting the `_datadog.config` object (before `web.wrapMiddleware` would get invoked), we can just reference `req._datadog.config.middleware` within `web.wrapMiddleware` without having to go through all the hassle of passing down config into it explicitly. Does that make sense?

File: packages/datadog-plugin-express/test/index.spec.js (hunk 6 above)
Comment (ericmustin) [reply packages/datadog-plugin-express/test/index.spec.js]: yup that's correct, i've update the language here to be more appropriate (which is now in line with how koa/connect tests for this are worded)

File: packages/dd-trace/src/plugins/util/web.js (hunk 8 above)
Comment (ericmustin) [reply packages/dd-trace/src/plugins/util/web.js]: I think I was using `reactivate` here inappropriately to try to do the work of both binding the  middleware function to the request span's scope, and invoking that middleware function. I've updated this to `tracer.scope().bind(fn, <request span>).apply(this, arguments)` and updated the method name to `bindAndWrapMiddlewareErrors` to more appropriately reflect the work being done here. I'm not 100% sure i follow where the ScopeManager would do that automatically, but if it's the case that I'm missing something here with how ScopeManager works and this call to `scope().bind` is unnecessary, please let me know

**Comment by ericmustin** [comment]
Not really clear to me why fs and pg are flaking? have to dig in a bit more, i dont think i touched anything on those?

**Comment by rochdev** [comment]
> Not really clear to me why fs and pg are flaking? have to dig in a bit more, i dont think i touched anything on those?

Build started to fail for those last week. It's unrelated to this PR.

File: packages/datadog-plugin-connect/src/index.js (hunk 4 above)
Comment (rochdev) [reply packages/datadog-plugin-connect/src/index.js]: Yes, and I'd say to definitely reuse the existing config.

File: packages/dd-trace/src/plugins/util/web.js (hunk 7 above)
Comment (rochdev) [reply packages/dd-trace/src/plugins/util/web.js]: The tracer handles errors explicitly, so it converts `name` to `type` and `message` to `msg`. This is only true for actual error objects however, but it should be the case here. If it doesn't work I'd say don't worry about it, but it probably means we have a bug in the tagger.

File: packages/datadog-plugin-express/test/index.spec.js [hunk 9]
```diff
@@ -720,7 +720,6 @@ describe('Plugin', () => {
           getPort().then(port => {
             agent.use(traces => {
               const spans = sort(traces[0])
-
```
Comment (rochdev) [inline packages/datadog-plugin-express/test/index.spec.js]: Unrelated change.

File: packages/datadog-plugin-koa/test/index.spec.js [hunk 10]
```diff
@@ -501,6 +501,329 @@ describe('Plugin', () => {
...
+                .catch(done)
+
+              appListener = app.listen(port, 'localhost', (e) => {
+                axios
+                  .get(`http://localhost:${port}/user/123`)
+                  .catch(done)
+              })
+            })
+
+            it('should support nested routers', done => {
```
Comment (rochdev) [inline packages/datadog-plugin-koa/test/index.spec.js]: There are some of these tests like this one that I'm not sure are relevant to the `middleware` option. In general I would expect a test that checks that no middleware span is created, and a test to check that the request scope is reactivated, and that's it. Maybe with the addition of testing error throwing, but even that I would expect to just work sine the scope is reactivated.

Can you clarify why the additional tests?

File: packages/dd-trace/src/plugins/util/web.js (hunk 7 above)
Comment (rochdev) [reply packages/dd-trace/src/plugins/util/web.js]: I also still don't think it's needed since `bind()` already adds the error to the span if the function throws.

Comment (ericmustin) [reply packages/dd-trace/src/plugins/util/web.js]: alrighty i figured out what was going on here regarding the `type` and `name` stuff not working. Basically, yes, `bind` or `setTag('error', e)` were working as intended, however there was a bit of odd behavior from  [web.addStatusError](https://github.com/DataDog/dd-trace-js/blob/5348eaad6f1a86d57b87ffd3de6397feb1dcb417/packages/dd-trace/src/plugins/util/web.js#L170) where it was basically defaulting (or in this case, overwriting) the `error` metadata to `true` unless the `req._datadog.error` key had already been explicitly set. I think this is _probably_ the correct behavior, or is a bit of a rabbit hole to rewrite at this point if it isn't, and so I've just added an explicit call to `web.addError` here in the catch block (as opposed to not having a catch block at all or having a catch block which called span.setTag), to set this `req._datadog.error` key explicitly. seems like a reasonable solution to me but lmk if i've missed something here.

File: packages/datadog-plugin-express/test/index.spec.js (hunk 9 above)
Comment (ericmustin) [reply packages/datadog-plugin-express/test/index.spec.js]: updated

File: packages/datadog-plugin-express/test/index.spec.js [hunk 11]
```diff
@@ -986,6 +985,207 @@ describe('Plugin', () => {
...
+          if (process.env.DD_CONTEXT_PROPAGATION === 'false') return done()
+
+          const app = express()
+
+          let span
+
+          app.use((req, res, next) => {
+            span = tracer.scope().active()
+            tracer.scope().activate(null, () => next())
+            next()
```
[the remark this answers is not in the document]
Comment (ericmustin) [reply packages/datadog-plugin-express/test/index.spec.js]: noted, not sure how that snuck in there, updated (tried to make sure i didnt see it anywhere else)

File: packages/datadog-plugin-koa/test/index.spec.js (hunk 10 above)
Comment (ericmustin) [reply packages/datadog-plugin-koa/test/index.spec.js]: when i first started this feature there was some additional complexity to the Koa router instrumentation that I didnt quite grok and just wanted to be sure wasn't getting messed with unintentionally. Ive reviewed these again and don't really think they're related at this point, so i agree and have removed these, although I did keep the 1 error handling test in there since there's a fair bit of this middleware code that overlaps with setting error metadata on the request span.

**Comment by ericmustin** [comment]
@rochdev 46th time's a charm 😅 , updated with all the last round of feedback

File: packages/dd-trace/src/plugins/util/web.js [hunk 12]
```diff
@@ -102,6 +107,16 @@ const web = {
     return tracer.scope().activate(span, fn)
   },
 
+  // catch errors and apply to active span
+  bindAndWrapMiddlewareErrors (fn, req, tracer, activeSpan) {
+    try {
+      return tracer.scope().bind(fn, activeSpan).apply(this, arguments)
+    } catch (e) {
+      web.addError(req, e)
```
Comment (rochdev) [inline packages/dd-trace/src/plugins/util/web.js]:
```suggestion
      web.addError(req, e) // TODO: remove when error formatting is moved to Span
```

**Comment by countcain** [comment]
@ericmustin Nice work. I just want to say thanks to you!
